### PR TITLE
New extended support for T2 

### PR DESCRIPTION
**Pull request by hkhlaaf**
We have extended llvm2kittel to better handle the T2 extension. This replaces the current version which only produces T2 files acceptable to run for termination. Now the produced files can run for all temporal logic handled by T2.

### CODE REVIEW

File: README.markdown [hunk 1]
```diff
@@ -4,9 +4,20 @@ llvm2KITTeL is a converter from LLVM's intermediate representation
 into a format that can be handled by the automatic termination prover
 [KITTeL](https://github.com/s-falke/kittel-koat).
 
-## Author
+This fork is a further extension to translate the LLVM-IR into the
```
Comment (s-falke) [inline README.markdown]: After a merge, this will no longer be a fork.  So please reword accordingly.

File: include/llvm2kittel/Converter.h [hunk 2]
```diff
@@ -138,7 +139,8 @@ class Converter : public llvm::InstVisitor<Converter>
 
     std::list<llvm::BasicBlock*> m_returns;
     std::map<llvm::Instruction*, unsigned int> m_idMap;
-
+    std::map<llvm::BasicBlock*, std::list<std::pair<std::string,llvm::Value*> > > m_phiMap;
```
Comment (s-falke) [inline include/llvm2kittel/Converter.h]: Please add " " between "," and "llvm::Value*"

File: lib/Core/Converter.cpp [hunk 3]
```diff
@@ -134,7 +138,7 @@ void Converter::phase1(llvm::Function *function, std::set<llvm::Function*> &scc,
     m_nondef = 0;
     m_entryBlock = &function->getEntryBlock();
     m_trivial = isTrivial();
-
+
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please remove trailing whitespaces.

File: lib/Core/Converter.cpp [hunk 4]
```diff
@@ -156,10 +160,11 @@ void Converter::phase1(llvm::Function *function, std::set<llvm::Function*> &scc,
         m_phase1 = true;
         visit(function);
     }
-
+
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please remove trailing whitespaces.

File: lib/Core/Converter.cpp [hunk 5]
```diff
@@ -156,10 +160,11 @@ void Converter::phase1(llvm::Function *function, std::set<llvm::Function*> &scc,
         m_phase1 = true;
         visit(function);
     }
-
+    
     for (std::list<std::string>::iterator i = m_vars.begin(), e = m_vars.end(); i != e; ++i) {
         m_lhs.push_back(Polynomial::create(*i));
     }
+
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please remove.

File: lib/Core/Converter.cpp [hunk 6]
```diff
@@ -245,8 +250,9 @@ std::list<ref<Rule> > Converter::getCondensedRules()
 std::string Converter::getVar(llvm::Value *V)
 {
     std::string name = V->getName();
+    name = "v" + name;
     std::ostringstream tmp;
-    tmp << "v_" << name;
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please make this change dependent on t2output.  I am not completely sure, but some other parts of llvm2kittel may rely on the naming scheme.

File: lib/Core/Converter.cpp [hunk 7]
```diff
@@ -299,6 +305,7 @@ ref<Polynomial> Converter::getPolynomial(llvm::Value *V)
         mpz_clear(value);
         return res;
     } else if (llvm::isa<llvm::Instruction>(V) || llvm::isa<llvm::Argument>(V) || llvm::isa<llvm::GlobalVariable>(V)) {
+        std::string res = getVar(V);
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: This line does not seem to have any effect.

File: lib/Core/Converter.cpp [hunk 8]
```diff
@@ -659,7 +674,7 @@ void Converter::visitBB(llvm::BasicBlock *bb)
     // visit instructions
     m_blockRules.clear();
     visit(*bb);
-
+
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please remove trailing whitespaces.

File: lib/Core/Converter.cpp [hunk 9]
```diff
@@ -700,10 +715,10 @@ void Converter::visitBB(llvm::BasicBlock *bb)
         ref<Rule> rule = Rule::create(lhs, rhs, cond);
         m_rules.push_back(rule);
     }
-
+
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please remove trailing whitespaces.

File: lib/Core/Converter.cpp [hunk 10]
```diff
@@ -700,10 +715,10 @@ void Converter::visitBB(llvm::BasicBlock *bb)
         ref<Rule> rule = Rule::create(lhs, rhs, cond);
         m_rules.push_back(rule);
     }
-
+    
     // insert rules for instructions
     m_rules.insert(m_rules.end(), m_blockRules.begin(), m_blockRules.end());
-
+
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please remove trailing whitespaces.

File: lib/Core/Converter.cpp [hunk 11]
```diff
@@ -722,7 +737,7 @@ void Converter::visitBB(llvm::BasicBlock *bb)
         ref<Rule> rule = Rule::create(lhs, rhs, Constraint::_true);
         m_rules.push_back(rule);
     }
-
+
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please remove trailing whitespaces.

File: lib/Core/Converter.cpp [hunk 12]
```diff
@@ -779,8 +797,116 @@ std::list<ref<Polynomial> > Converter::getArgsWithPhis(llvm::BasicBlock *from, l
     return res;
 }
 
-void Converter::visitTerminatorInst(llvm::TerminatorInst&)
-{}
+void Converter::visitTerminatorInst(llvm::TerminatorInst &I)
+{
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please wrap the whole body if an "if (m_t2Output)" block.

File: lib/Core/Converter.cpp [hunk 13]
```diff
@@ -1636,6 +1799,7 @@ void Converter::visitSelectInst(llvm::SelectInst &I)
         if (m_selectIsControl) {
             m_controlPoints.insert(getEval(m_counter));
         }
+        llvm::BasicBlock* pBlock = I.getParent();
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please sink this into the "if (m_t2Output)" block.

File: lib/Core/Converter.cpp [hunk 14]
```diff
@@ -1510,6 +1661,12 @@ void Converter::visitXor(llvm::BinaryOperator &I)
 
 void Converter::visitCallInst(llvm::CallInst &I)
 {
+    llvm::CallSite callSite(&I);
+    llvm::Function *calledFunction = callSite.getCalledFunction();
+    if (calledFunction != NULL) {
+        llvm::StringRef functionName = calledFunction->getName();
+    }
+
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Are any of the lines 1664-1669 needed/used?

File: lib/Core/Converter.cpp [hunk 15]
```diff
@@ -1658,7 +1842,22 @@ void Converter::visitPHINode(llvm::PHINode &I)
         std::string phiVar = getVar(&I);
         m_vars.push_back(phiVar);
         m_phiVars.insert(phiVar);
-    } else {
+        std::string valName;
+        unsigned incomeVals = I.getNumIncomingValues();
+        for(unsigned i = 0; i < incomeVals; i++)
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please add  a space between "for" and "("

File: lib/Core/Converter.cpp [hunk 16]
```diff
@@ -1658,7 +1842,22 @@ void Converter::visitPHINode(llvm::PHINode &I)
...
+        {
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please move to previous line.

File: lib/Core/Converter.cpp [hunk 17]
```diff
@@ -1658,7 +1842,22 @@ void Converter::visitPHINode(llvm::PHINode &I)
         std::string phiVar = getVar(&I);
         m_vars.push_back(phiVar);
         m_phiVars.insert(phiVar);
-    } else {
+        std::string valName;
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Used?

File: lib/Core/Converter.cpp [hunk 18]
```diff
@@ -1658,7 +1842,22 @@ void Converter::visitPHINode(llvm::PHINode &I)
...
+        unsigned incomeVals = I.getNumIncomingValues();
+        for(unsigned i = 0; i < incomeVals; i++)
+        {
+            llvm::BasicBlock* iBlock = I.getIncomingBlock(i);
+            llvm::Value* iValue = I.getIncomingValueForBlock(iBlock);
+            std::map<llvm::BasicBlock*, std::list<std::pair<std::string,llvm::Value*> > >::iterator it = m_phiMap.find(iBlock);
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please add space between "," and "llvm::Value*"

File: lib/Core/Converter.cpp [hunk 19]
```diff
@@ -1658,7 +1842,22 @@ void Converter::visitPHINode(llvm::PHINode &I)
...
+            if (it != m_phiMap.end()){
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please add space between ")" and "{"

File: lib/Core/Converter.cpp [hunk 20]
```diff
@@ -1658,7 +1842,22 @@ void Converter::visitPHINode(llvm::PHINode &I)
...
+                it->second.push_back(std::make_pair(phiVar,iValue));
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please add space between "," and "iValue"

File: lib/Core/Converter.cpp [hunk 21]
```diff
@@ -1658,7 +1842,22 @@ void Converter::visitPHINode(llvm::PHINode &I)
...
+            }
+            else{
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please move to previous line and add space between "else" and "{"

File: lib/Core/Converter.cpp [hunk 22]
```diff
@@ -1658,7 +1842,22 @@ void Converter::visitPHINode(llvm::PHINode &I)
...
+                std::list<std::pair<std::string,llvm::Value*> > valList;
+                valList.push_back(std::make_pair(phiVar,iValue));
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please add space between "," and "iValue"

File: lib/Core/Converter.cpp [hunk 23]
```diff
@@ -1753,12 +1956,18 @@ void Converter::visitStoreInst(llvm::StoreInst &I)
...
+        if (m_t2Output) {
+            //std::cout << (nondef->toString()) << ":= nondet();" << std::endl;
+            std::cout << (getVar(&I)) << " := nondet();" << std::endl;
+        }
         m_blockRules.push_back(rule);
     }
 }
 
 void Converter::visitAllocaInst(llvm::AllocaInst &)
-{}
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please keep it like this.

File: lib/Core/Converter.cpp [hunk 24]
```diff
@@ -1787,6 +1996,7 @@ void Converter::visitInstruction(llvm::Instruction &I)
     }
     if (m_phase1) {
         m_vars.push_back(getVar(&I));
+
```
Comment (s-falke) [inline lib/Core/Converter.cpp]: Please revert this change.

File: llvm2kittel-build/phiTest.c [hunk 25]
```diff
@@ -0,0 +1,9 @@
+//
+//  phiTest.c
+//  
+//
+//  Created by Heidy Khlaaf on 24/09/2015.
+//
+//
+
+#include <stdio.h>
```
Comment (s-falke) [inline llvm2kittel-build/phiTest.c]: I am not sure I understand the purpose of this file (and directory).

Comment (hkhlaaf) [inline llvm2kittel-build/phiTest.c]: Ah, this was added when I was testing it with Marc. It is not necessary and I will remove this.

**Comment by s-falke** [comment]
Thanks for the pull request!  I'll happily merge it, but I have some (mostly stylistic) suggestions/requests.  Please see the comments above.
